### PR TITLE
Add UDP support

### PR DESCRIPTION
**Pull request by leo-lb**

### CODE REVIEW

File: src/udp.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        Ok(UdpSocket {
+            registry: poll_registry.try_clone()?,
+            waker,
+            socket,
+            token,
+        })
+    }
+
+    pub fn connect(&self, addr: SocketAddr, poll_registry: &PollRegistry) -> IoResult<()> {
+        MioUdpSocket::connect(&self.socket, addr)
```
Comment (khionu) [inline src/udp.rs]: This needs to be registered with the PollRegistry here

Comment (khionu) [reply src/udp.rs]: Actually, sorry, hold on that, I was still thinking in TCP terms

File: src/udp.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<FutIoResult<(usize, SocketAddr)>> {
+        self.waker.get_read_waker().register(cx.waker());
+
+        match self.socket.recv_from(buf) {
+            Ok(len) => Ok(len).into(),
+            Err(ref err) if err.kind() == ErrorKind::WouldBlock => Poll::Pending,
+            Err(ref err) if err.kind() == ErrorKind::Interrupted => {
+                { cx.waker().clone() }.wake();
```
Comment (khionu) [inline src/udp.rs]: What's the purpose of putting this in brackets?

File: src/udp.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+};
+
+use futures::{future::poll_fn, task::AtomicWaker, AsyncRead, AsyncWrite, Stream};
+use log::{debug, trace};
+use mio::net::UdpSocket as MioUdpSocket;
+use mio::Interest;
+
+use crate::{FutIoResult, PollRegistry, SourceWaker, Token};
+
+pub struct UdpSocket {
```
Comment (khionu) [inline src/udp.rs]: This needs to be Clone so Read and Write can be done from separate threads. Will just need to make the `socket` Arc'd.

File: src/udp.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    registry: PollRegistry,
```
Comment (khionu) [inline src/udp.rs]: We don't need to hold onto this.

File: src/udp.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            socket,
+            token,
+        })
+    }
+
+    pub fn connect(&self, addr: SocketAddr) -> IoResult<()> {
+        MioUdpSocket::connect(&self.socket, addr)
+    }
+
+    pub fn poll_recv_from(
```
Comment (khionu) [inline src/udp.rs]: This should be private, as an internal function.

File: src/udp.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            }
+            err => err.into(),
+        }
+    }
+
+    pub async fn recv_from(&mut self, buf: &mut [u8]) -> IoResult<(usize, SocketAddr)> {
+        poll_fn(|cx| self.poll_recv_from(cx, buf)).await
+    }
+
+    pub fn poll_send_to(
```
Comment (khionu) [inline src/udp.rs]: This should be private as well.

File: src/udp.rs (hunk 2 above)
Comment (leo-lb) [reply src/udp.rs]: There isnt one, that code already existed as is in the project, I took it from `tcp.rs`.

Comment (khionu) [reply src/udp.rs]: Huh.... I must have accidentally added it while I was doing the initial attempts at porting to 0.7. It's not needed, but I'll remove it in a clean up sweep before 0.1

Comment (khionu) [reply src/udp.rs]: Huh.... I must have accidentally added it while I was doing the initial attempts at porting to 0.7. It's not needed, but I'll remove it in a clean up sweep before 0.1

**Pull request merged**